examples/s3/getobject-client-encryption.go

```diff
@@ -0,0 +1,105 @@
+// +build ignore
+
+/*
+ * Minio Go Library for Amazon S3 Compatible Cloud Storage
+ * Copyright 2018 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"errors"
+	"io"
+	"log"
+	"os"
+	"path"
+
+	"github.com/minio/minio-go"
+	"github.com/minio/sio"
+	"golang.org/x/crypto/argon2"
+)
+
+const (
+	// SSE DARE package block size.
+	sseDAREPackageBlockSize = 64 * 1024 // 64KiB bytes
+
+	// SSE DARE package meta padding bytes.
+	sseDAREPackageMetaSize = 32 // 32 bytes
+)
+
+func decryptedSize(encryptedSize int64) (int64, error) {
+	if encryptedSize == 0 {
+		return encryptedSize, nil
+	}
+	size := (encryptedSize / (sseDAREPackageBlockSize + sseDAREPackageMetaSize)) * sseDAREPackageBlockSize
+	if mod := encryptedSize % (sseDAREPackageBlockSize + sseDAREPackageMetaSize); mod > 0 {
+		if mod < sseDAREPackageMetaSize+1 {
+			return -1, errors.New("object is tampered")
+		}
+		size += mod - sseDAREPackageMetaSize
+	}
+	return size, nil
+}
+
+func main() {
+	// Note: YOUR-ACCESSKEYID, YOUR-SECRETACCESSKEY, my-testfile, my-bucketname and
+	// my-objectname are dummy values, please replace them with original values.
+
+	// Requests are always secure (HTTPS) by default. Set secure=false to enable insecure (HTTP) access.
+	// This boolean value is the last argument for New().
+
+	// New returns an Amazon S3 compatible client object. API compatibility (v2 or v4) is automatically
+	// determined based on the Endpoint value.
+	s3Client, err := minio.New("s3.amazonaws.com", "YOUR-ACCESSKEYID", "YOUR-SECRETACCESSKEY", true)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	obj, err := s3Client.GetObject("my-bucketname", "my-objectname", minio.GetObjectOptions{})
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	objSt, err := obj.Stat()
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	size, err := decryptedSize(objSt.Size)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	localFile, err := os.Create("my-testfile")
+	if err != nil {
+		log.Fatalln(err)
+	}
+	defer localFile.Close()
+
+	password := "myfavoritepassword"                    // Change as per your needs.
+	salt := path.Join("my-bucketname", "my-objectname") // Change as per your needs.
+	decrypted, err := sio.DecryptReader(obj, sio.Config{
+		Key: argon2.IDKey(password, salt, 1, sseDAREPackageBlockSize, 4, sseDAREPackageMetaSize),
```

Comment:
You can use `encrypt.DefaultPBKDF` here.

Reply:
in both examples the argon2.IDKey method needs to have password and salt converted to byte array
```
encrypted, err := sio.EncryptReader(object, sio.Config{
		Key: argon2.IDKey([]byte(password), []byte(salt), 1, sseDAREPackageBlockSize, 4, sseDAREPackageMetaSize),
	})
decrypted, err := sio.DecryptReader(obj, sio.Config{
		Key: argon2.IDKey([]byte(password), []byte(salt), 1, sseDAREPackageBlockSize, 4, sseDAREPackageMetaSize),
	})
```

Reply:
> You can use encrypt.DefaultPBKDF here.

I don't want to use code from server-side.go which has special methods attached to it - which will confuse the user @aead - here PBKDF is all about client side and how they can leverage it using argon2.

Reply:
updated @poornas

Reply:
@harshavardhana Okay - than one last thing. The parameters for argon2 (`sseDAREPackageBlockSize`, `sseDAREPackageMetaSize`) look like they are somehow tied to DARE / sio - but they are not. 64 MB is a (recommended) value from the RFC draft and the 32 is the key size. I see you are using constants here instead of literals but IMO the constants give a wrong impression of the meaning of these values. Maybe we can use literals or properly named constants. (I didn't mentioned it before because of `PBKDF` suggestion)